lmdeploy/pytorch/backends/default/token_dispatcher.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Tuple
+
+import torch
+
+from ..token_dispatcher import TokenDispatcherImpl
+
+
+class AlltoAllTokenDispatcher(TokenDispatcherImpl):
+
+    def __init__(
+        self,
+        ep_group,
+        num_experts,
+        num_local_experts: int,
+    ) -> None:
+        self.num_local_experts = num_local_experts
+        assert num_experts is not None
+        self.num_experts = num_experts
+        assert self.num_local_experts > 0, 'Expected at least one expert'
+        self.ep_size = num_experts // num_local_experts
+        self.ep_group = ep_group
+        self.tp_size = 1
+        self.input_splits = None
+        self.output_splits = None
+        self.permute_idx_device = torch.device('cuda')
+        input_chunk_idxs = torch.arange(self.num_experts, device=self.permute_idx_device)
+        self.sort_input_by_local_experts = input_chunk_idxs.reshape(-1, self.num_local_experts).T.ravel()
+        self.restore_output_by_local_experts = input_chunk_idxs.reshape(self.num_local_experts, -1).T.ravel()
+
+    def sort_chunks_by_idxs(self,
+                            input: torch.Tensor,
+                            split_sizes: torch.Tensor,
+                            sorted_idxs: torch.Tensor,
+                            fused: bool = False):
+        """Split and sort the input tensor based on the split_sizes and sorted
+        indices."""
+        input = torch.split(input, split_sizes.tolist(), dim=0)
+        output = torch.cat([input[i] for i in sorted_idxs.tolist()], dim=0)
+        return output
+
+    def all_to_all(self, group, input_, output_split_sizes_=None, input_split_sizes=None):
+        output = input_.new_empty(
+            size=[sum(output_split_sizes_)] + list(input_.size()[1:]),
+            dtype=input_.dtype,
+            device=torch.cuda.current_device(),
+        )
+        torch.distributed.all_to_all_single(
+            output,
+            input_,
+            output_split_sizes=output_split_sizes_,
+            input_split_sizes=input_split_sizes,
+            group=group,
+        )
+        return output
+
+    def preprocess(self, routing_map: torch.Tensor, local_expert_indices) -> torch.Tensor:
+        assert (len(local_expert_indices) == self.num_local_experts), 'Invalid local expert indices'
+        for i in range(len(local_expert_indices) - 1):
+            assert (local_expert_indices[i] == local_expert_indices[i + 1] -
+                    1), 'local_expert_indices must be continous'
+
+        num_local_tokens_per_expert = routing_map.sum(dim=0).long()
+        self.input_splits = (num_local_tokens_per_expert.reshape(self.ep_size, self.num_local_experts).sum(axis=1).to(
+            torch.device('cpu'), non_blocking=True).numpy())
```

Comment:
Since you use `non_blocking=True` here, why not delay `numpy()` when using `input_splits`?